src/lib.rs

```diff
@@ -1,39 +1,84 @@
 
-mod receiver;
-pub use receiver::Receiver;
+/// This is a convenience alias where the lifetime is static.
+pub type Receiver<T> = receiver::Receiver<'static, T>;
+/// This is a convenience alias where the lifetime is static.
+pub type Sender<T> = sender::Sender<'static, T>;
+
+#[cfg(feature="alloc")]
+/// Creates a new hatch backed by a box. Returns a [`Sender`]/[`Receiver`] pair.
+pub fn hatch<T>() -> (Sender<T>, Receiver<T>) {
+    let unbox = Box::leak(Box::new(Hatch::default()));
+    let non = unsafe { NonNull::new_unchecked(unbox) };
+    let holder = Holder::SharedBoxPtr(non);
+    // Safe because we have exclusive access
+    unsafe { (Sender::new(holder), Receiver::new(holder)) }
+}
 
-/// Create a new oneshot channel pair.
+#[cfg(feature="alloc")]
+/// Like [`hatch`], but configured to close after the first message. Returns a [`Sender`]/[`Receiver`] pair.
 pub fn oneshot<T>() -> (Sender<T>, Receiver<T>) {
-    let inner = Arc::new(Inner::new());
-    let sender = Sender::new(inner.clone());
-    let receiver = Receiver::new(inner);
-    (sender, receiver)
+    let (s, r) = hatch();
+    (s.close_on_send(true), r.close_on_receive(true))
+}
+
+/// Creates a new hatch backed by a ref to an existing Hatch. Unlike
+/// [`hatch`], this is not allocated on the stack and is bound by the
+/// lifetime of the passed mut ref.
+pub fn ref_hatch<T>(
+    hatch: Pin<&mut Hatch<T>>
+) -> (sender::Sender<'_, T>, receiver::Receiver<'_, T>) {
+    // Safe because we aren't going to move out of it
+    let hatch = unsafe { Pin::into_inner_unchecked(hatch) };
+    let holder = Holder::Ref(hatch);
+    // Safe because we have exclusive access
+    unsafe { (sender::Sender::new(holder), receiver::Receiver::new(holder)) }
+}
+
+/// Like [`ref_hatch`], but takes an immutable ref.
+///
+/// ## Safety
+///
+/// You must not permit multiple live senders or receivers to exist.
+pub unsafe fn ref_hatch_unchecked<T>(
+    hatch: &Hatch<T>
+) -> (sender::Sender<T>, receiver::Receiver<T>) {
+    let holder = Holder::Ref(hatch);
+    (sender::Sender::new(holder), receiver::Receiver::new(holder))
 }
 
-/// An empty struct that signifies the channel is closed.
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct Closed();
-
-/// We couldn't receive a message.
-#[derive(Debug)]
-pub enum TryRecvError<T> {
-    /// The Sender didn't send us a message yet.
-    Empty(Receiver<T>),
-    /// The Sender has dropped.
-    Closed,
+/// Creates a new hatch backed by a `core::ptr::NonNull` to an
+/// existing [`Hatch`] whose lifetime is asserted to be whatever
+/// lifetime you say it is.
+///
+/// # Safety
+///
+/// You must not permit multiple live senders or receivers to exist.
+pub unsafe fn borrowed_ptr_hatch<'a, T>(
+    hatch: NonNull<Hatch<T>>
+) -> (sender::Sender<'a, T>, receiver::Receiver<'a, T>) {
+    let holder = Holder::BorrowedPtr(hatch);
+    // Safe because we have exclusive access
+    (sender::Sender::new(holder), receiver::Receiver::new(holder))
 }
```

Comment:
I think this shouldn't be in the interface at all. It is incredible unsafe, and I can't imagine a scenario where a usecase fulfilled by this isn't also fulfilled with `ref_hatch_unchecked(std::mem::transmute::<&Hatch<T>, &'a Hatch<T>>(hatch))`.

Reply:
`ref_hatch_unchecked` is unsafe because without a mutable reference, we can't guarantee there are no other senders or receivers and without a pin, we can't guarantee it won't move. while the same can be said of `borrowed_ptr_hatch`, i'm not sure why you think it's worse? that said, it existed first and i haven't really evaluated whether it's still required.

a transmute is not required, you can cast through a pointer to achieve the same effect.